File: src/PythonUnitTests/TwoDimBaseTests.py
import unittest
import numpy as np
from nptest import nptest

class Test_test1(unittest.TestCase):
  
    def test_diag_1(self):
        m = np.arange(9);
        n = np.diag(m)

        print(m)
        print(n)

        m = np.arange(9).reshape(3,3);
        n = np.diag(m)

        print(m)
        print(n)

    def test_diagflat_1(self):
        m = np.arange(1,5).reshape(2,2);
        n = np.diagflat(m)

        print(m)
        print(n)

        m = np.arange(1,3)
        n = np.diagflat(m, 1)

        print(m)
        print(n)

        m = np.arange(1,3)
        n = np.diagflat(m, -1)

        print(m)
        print(n)

        
    def test_eye_1(self):

        a = np.eye(2, dtype = np.int32)

        print(a)
        print(a.shape)
        print(a.strides)
        print("")

        b = np.eye(3, k = 1)

        print(b)
        print(b.shape)
        print(b.strides)

        
    def test_fliplr_1(self):
        m = np.arange(8).reshape((2,2,2))
        n = np.fliplr(m)

        print(m)
        print(n)

    def test_flipud_1(self):
        m = np.arange(8).reshape((2,2,2))
        n = np.flipud(m)

        print(m)
        print(n)


    def test_tri_1(self):
        print(np.tri(3, 5, 2, dtype=int))
        print("***********")
        print(np.tri(3, 5, -1))


    def test_tril_1(self):

        a = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])
        b = np.tril(a, -1)
        print(a)
        print("***********")
        print(b)
  
    def test_triu_1(self):

        a = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])
        b = np.triu(a, -1)
        print(a)
        print("***********")
        print(b)

    def test_vander_1(self):

        x = np.array([1, 2, 3, 5])
        N = 3

        v = np.empty((len(x), N), dtype=np.promote_types(x.dtype, int))

        y = np.vander(x, N)
        print(y)

        y = np.vander(x)
        print(y)

        y = np.vander(x, increasing=True)
        print(y)

        return
    
    def test_histogram2d(self):

        return
    
    def test_mask_indices(self):

        iu = np.mask_indices(3, np.triu)
        print(iu)

        a = np.arange(9).reshape(3, 3)
        b = a[iu]
        print(b)

        iu1 = np.mask_indices(3, np.triu, 1)

        c = a[iu1]
        print(c)

        return

    
    def test_tril_indices(self):

        il1 = np.tril_indices(4)
        il2 = np.tril_indices(4, 2)

        a = np.arange(16).reshape(4, 4)
        b = a[il1]
        print(b)

        a[il1] = -1
        print(a)
 
        a[il2] = -10
        print(a)
 
        return
    
    def test_tril_indices_from(self):

        a = np.arange(16).reshape(4, 4)
        il1 = np.tril_indices_from(a, 0)
        il2 = np.tril_indices_from(a, 2)

        print(il1)
        print(il2)

        return
    
    def test_triu_indices(self):

        il1 = np.triu_indices(4)
        il2 = np.triu_indices(4, 2)

        a = np.arange(16).reshape(4, 4)
        b = a[il1]
        print(b)

        a[il1] = -1
        print(a)
 
        a[il2] = -10
        print(a)
 
        return
    
    def test_triu_indices_from(self):

        a = np.arange(16).reshape(4, 4)
        il1 = np.triu_indices_from(a, 0)
        il2 = np.triu_indices_from(a, 2)

        print(il1)
        print(il2)

        return

if __name__ == '__main__':
    unittest.main()

File: src/NumpyDotNet/UnitTests/NumpyDotNetTests/TwoDimBaseTests.cs
using System;
using System.Numerics;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using NumpyDotNet;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using NumpyLib;

namespace NumpyDotNetTests
{
    [TestClass]
    public class TwoDimBaseTests : TestBaseClass
    {
        [TestMethod]
        public void test_diag_1()
        {
            ndarray m = np.arange(9);
            var n = np.diag(m);

            print(m);
            print(n);

            var ExpectedDataN = new Int32[,]
                {{0, 0, 0, 0, 0, 0, 0, 0, 0},
                 {0, 1, 0, 0, 0, 0, 0, 0, 0},
                 {0, 0, 2, 0, 0, 0, 0, 0, 0},
                 {0, 0, 0, 3, 0, 0, 0, 0, 0},
                 {0, 0, 0, 0, 4, 0, 0, 0, 0},
                 {0, 0, 0, 0, 0, 5, 0, 0, 0},
                 {0, 0, 0, 0, 0, 0, 6, 0, 0},
                 {0, 0, 0, 0, 0, 0, 0, 7, 0},
                 {0, 0, 0, 0, 0, 0, 0, 0, 8}};

            AssertArray(n, ExpectedDataN);

            m = np.arange(9).reshape(new shape(3, 3));
            n = np.diag(m);

            print(m);
            print(n);
            AssertArray(n, new int[] { 0, 4, 8 });
        }

        [TestMethod]
        public void test_diagflat_1()
        {
            ndarray m = np.arange(1, 5).reshape(new shape(2, 2));
            var n = np.diagflat(m);

            print(m);
            print(n);

            var ExpectedDataN = new Int32[,]
            {
             {1, 0, 0, 0},
             {0, 2, 0, 0},
             {0, 0, 3, 0},
             {0, 0, 0, 4}
            };
            AssertArray(n, ExpectedDataN);

            m = np.arange(1, 3);
            n = np.diagflat(m, 1);

            print(m);
            print(n);

            ExpectedDataN = new Int32[,]
            {
             {0, 1, 0},
             {0, 0, 2},
             {0, 0, 0},
            };

            AssertArray(n, ExpectedDataN);

            m = np.arange(1, 3);
            n = np.diagflat(m, -1);

            print(m);
            print(n);

            ExpectedDataN = new Int32[,]
            {
             {0, 0, 0},
             {1, 0, 0},
             {0, 2, 0},
            };

            AssertArray(n, ExpectedDataN);

        }


        [TestMethod]
        public void test_eye_1()
        {
            ndarray a = np.eye(2, dtype: np.Int32);

            print(a);
            print(a.shape);
            print(a.strides);

            var ExpectedDataA = new Int32[2, 2]
            {
                { 1,0 },
                { 0,1 },
            };
            AssertArray(a, ExpectedDataA);
            AssertShape(a, 2, 2);
            AssertStrides(a, 8, 4);


            ndarray b = np.eye(3, k: 1);

            print(b);
            print(b.shape);
            print(b.strides);

            var ExpectedDataB = new double[3, 3]
            {
                { 0,1,0 },
                { 0,0,1 },
                { 0,0,0 },
            };

            AssertArray(b, ExpectedDataB);
            AssertShape(b, 3, 3);
            AssertStrides(b, 24, 8);


        }


        [TestMethod]
        public void test_fliplr_1()
        {
            ndarray m = np.arange(8).reshape(new shape(2, 2, 2));
            var n = np.fliplr(m);

            print(m);
            print(n);

            AssertArray(n, new Int32[,,] { { { 2, 3 }, { 0, 1 } }, { { 6, 7 }, { 4, 5 } } });
        }

        [TestMethod]
        public void test_flipud_1()
        {
            ndarray m = np.arange(8).reshape(new shape(2, 2, 2));
            var n = np.flipud(m);

            print(m);
            print(n);

            AssertArray(n, new Int32[,,] { { { 4, 5 }, { 6, 7 } }, { { 0, 1 }, { 2, 3 } } });
        }



        [TestMethod]
        public void test_tri_1()
        {
            ndarray a = np.tri(3, 5, 2, dtype: np.Int32);
            print(a);

            var ExpectedDataA = new Int32[,]
            {
             {1, 1, 1, 0, 0},
             {1, 1, 1, 1, 0},
             {1, 1, 1, 1, 1}
            };
            AssertArray(a, ExpectedDataA);

            print("***********");
            ndarray b = np.tri(3, 5, -1);
            print(b);

            var ExpectedDataB = new float[,]
            {
             {0.0f, 0.0f, 0.0f, 0.0f, 0.0f},
             {1.0f, 0.0f, 0.0f, 0.0f, 0.0f},
             {1.0f, 1.0f, 0.0f, 0.0f, 0.0f}
            };
            AssertArray(b, ExpectedDataB);
        }

        [TestMethod]
        public void test_tril_1()
        {
            ndarray a = np.array(new int[,] { { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 }, { 10, 11, 12 } });
            ndarray b = np.tril(a, -1);
            print(a);
            print("***********");
            print(b);

            var ExpectedDataB = new Int32[,]
            {
             {0, 0, 0},
             {4, 0, 0},
             {7, 8, 0},
             {10, 11, 12},
            };
            AssertArray(b, ExpectedDataB);

        }

        [TestMethod]
        public void test_triu_1()
        {
            ndarray a = np.array(new int[,] { { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 }, { 10, 11, 12 } });
            ndarray b = np.triu(a, -1);
            print(a);
            print("***********");
            print(b);

            var ExpectedDataB = new Int32[,]
            {
             {1, 2, 3},
             {4, 5, 6},
             {0, 8, 9},
             {0, 0, 12},
            };
            AssertArray(b, ExpectedDataB);

        }


        [Ignore] // not implemented yet
        [TestMethod]
        public void test_vander_1()
        {

        }


        [Ignore] // not implemented yet
        [TestMethod]
        public void test_histogram2d()
        {

        }

        [TestMethod]
        public void test_mask_indices()
        {
            var iu = np.mask_indices(3, np.triu);
            AssertArray(iu[0], new Int64[] { 0, 0, 0, 1, 1, 2 });
            AssertArray(iu[1], new Int64[] { 0, 1, 2, 1, 2, 2 });
            print(iu);

            var a = np.arange(9).reshape((3, 3));
            var b = a[iu] as ndarray;
            AssertArray(b, new Int64[] { 0, 1, 2, 4, 5, 8 });
            print(b);

            var iu1 = np.mask_indices(3, np.triu, 1);

            var c = a[iu1] as ndarray;
            AssertArray(c, new Int64[] { 1, 2, 5 });
            print(c);

            return;
        }

        [TestMethod]
        public void test_tril_indices()
        {
            var il1 = np.tril_indices(4);
            var il2 = np.tril_indices(4, 2);

            var a = np.arange(16).reshape((4, 4));
            var b = a[il1] as ndarray;
            AssertArray(b, new int[] { 0, 4, 5, 8, 9, 10, 12, 13, 14, 15 });
            print(b);

            a[il1] = -1;

            var ExpectedDataA1 = new int[,]
                {{-1,  1, 2,  3}, {-1, -1,  6,  7}, 
                 {-1, -1,-1, 11}, {-1, -1, -1, -1}};
            AssertArray(a, ExpectedDataA1);
            print(a);

            a[il2] = -10;

            var ExpectedDataA2 = new int[,]
                {{-10, -10, -10,  3}, {-10, -10, -10, -10},
                 {-10, -10,-10, -10}, {-10, -10, -10, -10}};
            AssertArray(a, ExpectedDataA2);
            print(a);

            return;
        }

        [Ignore] // not implemented yet
        [TestMethod]
        public void test_tril_indices_from()
        {

        }

        [Ignore] // not implemented yet
        [TestMethod]
        public void test_triu_indices()
        {

        }

        [Ignore] // not implemented yet
        [TestMethod]
        public void test_triu_indices_from()
        {

        }

    }
}
